Upload dataset from efs to wasabi and glacier

kencho51 opened the pull request:
# Pull request for issue: #1903, #1771, #1861

This is a pull request for the following functionalities:

1. Created a wrapper script `transfer.sh` to execute 2 rclone copy actions
2. `--wasabi` option will copy files from local to wasabi bucket
3. `--backup` option will copy files from local s3 bucket in glacier tier
4. These two options can be supplied separately or together 
5. Automatically mount aws efs access points `dropbox` and `configuration` to the bastion server
6. Automatically mount aws efs acces point `configuration` to the webapp server
7. Output the log to the `/var/log/gigadb` on the bastion server which can be accessed any users
8. Define the profile name `[aws-transfer]` and `[wasabi-transfer]` in the `/home/$user/~.aws/credentials`, by adding parameter [s3-profile](https://rclone.org/s3/#s3-profile) to the rclone cmd, the rclone will then able to get the corresponding secrets from `/home/$user/~.aws/credentials`

## How to test?

Describe how the new functionalities can be tested by PR reviewers

### Pre-requisite
1. rclone is installed in dev, eg:
```
 % rclone --version
rclone v1.67.0
- os/version: darwin 13.4.1 (64 bit)
- os/kernel: 22.5.0 (arm64)
- os/type: darwin
- os/arch: arm64 (ARMv8 compatible)
- go/version: go1.22.4
- go/linking: dynamic
- go/tags: none
```
2. The following variables are added into your gitlab project variable page

| key | value | env |
| --- | --- | --- |
| WASABI_ACCESS_KEY_ID | $developer_wasabi_access_key | dev     |
| WASABI_SECRET_ACCESS_KEY | $developer_wasabi_secret_access_key | dev     |
| WASABI_ACCESS_KEY_ID | $developer_wasabi_access_key | staging |
| WASABI_SECRET_ACCESS_KEY | $developer_wasabi_secret_access_key | staging |
| WASABI_ACCESS_KEY_ID | $developer_wasabi_access_key | live    |
| WASABI_SECRET_ACCESS_KEY | $developer_wasabi_secret_access_key | live    |
| WASABI_DATASETFILES_DIR | wasabi:gigadb-datasets/dev/pub/10.5524 | dev |
| WASABI_DATASETFILES_DIR | wasabi:gigadb-datasets/staging/pub/10.5524 | staging |
| WASABI_DATASETFILES_DIR | wasabi:gigadb-datasets/live/pub/10.5524 | live |
| S3_DATASETFILES_DIR| gigadb-datasetfiles:gigadb-datasetfiles-backup/dev/pub/10.5524 | dev |
| S3_DATASETFILES_DIR| gigadb-datasetfiles:gigadb-datasetfiles-backup/staging/pub/10.5524 | staging |
| S3_DATASETFILES_DIR| gigadb-datasetfiles:gigadb-datasetfiles-backup/live/pub/10.5524 | live |

3. Check that these variables have been added to `cnhk-infra` variable page

| key | value | env |
| --- | --- | --- |
| gigadb_datasetfiles_aws_access_key_id | xxxxxxxxxxxxxxxxxx | All (default) |
| gigadb_datasetfiles_aws_secret_access_key | yyyyyyyyyyyyyyyyy | All (default) |


### In dev
```
# update the rclone config file
% cd gigadb/app/tools/transfer-files
# execute the configure script to create the .env file
% ./configure
# in the new `.env` file, uncomment and provide a value for the `GITLAB_PRIVATE_TOKEN` variable and also fill in the REPO_NAME
# then execute the configure script again to create the .secrets file
# execute the configure script again to create the .secrets file
% ./configure
# Make sure the following credentials have been added into developer's `~/.aws/credentials`, the values from the `.secrets` file
% vi ~/.aws/credentials
[wasabi-transfer]
aws_access_key_id = $WASABI_ACCESS_KEY_ID
aws_secret_access_key = $WASABI_SECRET_ACCESS_KEY

[aws-transfer]
aws_access_key_id = $gigadb_datasetfiles_aws_access_key_id
aws_secret_access_key = $gigadb_datasetfiles_aws_secret_access_key
# execute the bats tests
% bats tests/bats/transfer.bats
transfer.bats
 ✓ No DOI provided
 ✓ Show error and usage if no flag
 ✓ Input DOI out of range
 ✓ Copy files from dev to Wasabi in dry run mode
 ✓ Copy files from dev to Wasabi with apply flag
 ✓ Copy files from dev to s3 in dry run mode
 ✓ Copy files from dev to s3 with apply flag
 ✓ Copy files from dev to Wasabi and s3 in dry run mode
 ✓ Copy files from dev to Wasabi and s3 and apply flag

9 tests, 0 failures


```


### Pre-requisite for ec2 servers provisioning

1. The wasabi user with group Curators assigned and the `credentials.csv`
2. Run `ansible-galaxy install -r ../../../infrastructure/requirements.yml` to install additional posix module for mounting access points
3. Follow the docs/SETUP_PROVISIONING.md to spin up servers
4. Make sure the gitlab pipeline is passing
6. Copy data to efs for testing:

```
% cd gigadb/app/tools/transfer-files
% scp -r -i /path/to/staging/pem tests/_data/102480 centos@$staging-bastion-ip:/share/dropbox/user101
Tree_file.txt                                                                                                                                                                                    100%  359     3.6KB/s   00:00    
readme_102480.txt                                                                                                                                                                                100% 3202    32.0KB/s   00:00  
% scp -r -i /path/to/live/pem tests/_data/102480 centos@$live-bastion-ip:/share/dropbox/user101
Tree_file.txt                                                                                                                                                                                    100%  359     3.6KB/s   00:00    
readme_102480.txt                                                                                                                                                                                100% 3202    32.0KB/s   00:00  
```
### As a centos user in staging
```
% ssh -i path/to/staging/pem centos@$staging-bastion-ip
Activate the web console with: systemctl enable --now cockpit.socket

Last login: Tue Jul  9 05:35:15 2024 from 3.36.204.163
# Confirm access points are mounted 
[centos@ip-10-99-0-151 ~]$ df -hT
Filesystem     Type      Size  Used Avail Use% Mounted on
devtmpfs       devtmpfs  339M     0  339M   0% /dev
tmpfs          tmpfs     372M     0  372M   0% /dev/shm
tmpfs          tmpfs     372M  9.9M  362M   3% /run
tmpfs          tmpfs     372M     0  372M   0% /sys/fs/cgroup
/dev/nvme0n1p1 xfs        30G  5.3G   25G  18% /
tmpfs          tmpfs      75M     0   75M   0% /run/user/1000
127.0.0.1:/    nfs4      8.0E     0  8.0E   0% /share/dropbox
127.0.0.1:/    nfs4      8.0E     0  8.0E   0% /share/config
tmpfs          tmpfs      75M     0   75M   0% /run/user/1001
[centos@ip-10-99-0-151 ~]$ ls -al /share/
t$ ls -al /share/
total 8
drwxr-xr-x.  4 centos centos   35 Jul 29 03:37 .
dr-xr-xr-x. 18 root   root    237 Jul 29 03:37 ..
drwxr-xr-x.  2 centos centos 6144 Jul 30 04:01 config
drwxr-xr-x.  2 centos centos 6144 Jul 30 04:01 dropbox

[centos@ip-10-99-0-240 ~]$ 
[centos@ip-10-99-0-212 ~]$ cat files-env 
GIGADB_ENV=staging
WASABI_DATASETFILES_DIR=wasabi:gigadb-datasets/staging/pub/10.5524
S3_DATASETFILES_DIR=gigadb-datasetfiles:gigadb-datasetfiles-backup/staging/pub/10.5524
[centos@ip-10-99-0-212 ~]$ ls /share/dropbox/user101/
analysis_data  readme_102480.txt
[centos@ip-10-99-0-212 ~]$ ls /share/dropbox/user101/analysis_data/
Tree_file.txt
[centos@ip-10-99-0-212 ~]$ /usr/local/bin/transfer
Usage: /usr/local/bin/transfer --doi <DOI> --sourcePath <Source Path>

Required:
--doi            Specify DOI to process
--sourcePath     Specify source path
--wasabi         Copy files to Wasabi bucket
--backup         Copy files to s3 bucket

Available Option:
--apply          Escape dry run mode

Example usages:
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi --apply
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --backup
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --backup --apply
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi --backup
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi --backup --apply
[centos@ip-10-99-0-212 ~]$ /usr/local/bin/transfer --doi 102480 --sourcePath /share/dropbox/user101/
Error: please specify --wasabi or --backup or both
Usage: /usr/local/bin/transfer --doi <DOI> --sourcePath <Source Path>

Required:
--doi            Specify DOI to process
--sourcePath     Specify source path
--wasabi         Copy files to Wasabi bucket
--backup         Copy files to s3 bucket

Available Option:
--apply          Escape dry run mode

Example usages:
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi --apply
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --backup
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --backup --apply
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi --backup
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi --backup --apply
[centos@ip-10-99-0-212 ~]$ 
[centos@ip-10-99-0-212 ~]$ /usr/local/bin/transfer --doi 102480 --sourcePath /share/dropbox/user101/ --wasabi --backup
[centos@ip-10-99-0-212 ~]$ ls /var/log/gigadb/
backup_20240805_041137.log
[centos@ip-10-99-0-212 ~]$ cat /var/log/gigadb/backup_20240805_041137.log 
2024/08/05 04:11:37 INFO  : Start copying files from staging to Wasabi
2024/08/05 04:11:37 NOTICE: analysis_data/Tree_file.txt: Skipped update modification time as --dry-run is set (size 359)
2024/08/05 04:11:37 NOTICE: readme_102480.txt: Skipped update modification time as --dry-run is set (size 3.127Ki)
2024/08/05 04:11:37 INFO  : There was nothing to transfer
2024/08/05 04:11:37 NOTICE: 
Transferred:              0 B / 0 B, -, 0 B/s, ETA -
Checks:                 2 / 2, 100%
Elapsed time:         0.5s

2024/08/05 04:11:37 INFO  : Executed: rclone copy --s3-no-check-bucket /share/dropbox/user101/ wasabi:gigadb-datasets/staging/pub/10.5524/102001_103000/102480 --dry-run --log-file /var/log/gigadb/backup_20240805_041137.log --log-level INFO --stats-log-level DEBUG >> /var/log/gigadb/backup_20240805_041137.log
2024/08/05 04:11:37 INFO  : Successfully copied files to Wasabi bucket for DOI: 102480
2024/08/05 04:11:37 INFO  : Start copying files from staging to s3
2024/08/05 04:11:38 NOTICE: readme_102480.txt: Skipped update modification time as --dry-run is set (size 3.127Ki)
2024/08/05 04:11:38 NOTICE: analysis_data/Tree_file.txt: Skipped update modification time as --dry-run is set (size 359)
2024/08/05 04:11:38 INFO  : There was nothing to transfer
2024/08/05 04:11:38 NOTICE: 
Transferred:              0 B / 0 B, -, 0 B/s, ETA -
Checks:                 2 / 2, 100%
Elapsed time:         0.4s

2024/08/05 04:11:38 INFO  : Executed: rclone copy --s3-no-check-bucket /share/dropbox/user101/ gigadb-datasetfiles:gigadb-datasetfiles-backup/staging/pub/10.5524/102001_103000/102480 --dry-run --log-file /var/log/gigadb/backup_20240805_041137.log --log-level INFO --stats-log-level DEBUG >> /var/log/gigadb/backup_20240805_041137.log
2024/08/05 04:11:38 INFO  : Successfully copied files to s3 bucket for DOI: 102480
[centos@ip-10-99-0-212 ~]$ 
```

### As a user lily in staging
```
% cd ops/infrastructure/envs/staging
% ansible-playbook -i ../../inventories users_playbook.yml -e "newuser=lily" -e "credentials_csv_path=~/path/to/credentials.csv" -e "gigadb_env=staging" 
% chmod 500 output/privkeys-$bastion-ip/lily
% ssh -i output/privkeys-3.36.204.163/lily lily@$bastion-ip
Activate the web console with: systemctl enable --now cockpit.socket

[lily@ip-10-99-0-212 ~]$ ls
files-env  uploadDir
# confirm the files-env contains staging variables
[lily@ip-10-99-0-212 ~]$ cat files-env 
GIGADB_ENV=staging
WASABI_DATASETFILES_DIR=wasabi:gigadb-datasets/staging/pub/10.5524
S3_DATASETFILES_DIR=gigadb-datasetfiles:gigadb-datasetfiles-backup/staging/pub/10.5524
[lily@ip-10-99-0-212 ~]$ ls /share/dropbox/user101/
analysis_data  readme_102480.txt
[lily@ip-10-99-0-212 ~]$ ls /share/dropbox/user101/analysis_data/
Tree_file.txt
[lily@ip-10-99-0-212 ~]$ /usr/local/bin/transfer
Usage: /usr/local/bin/transfer --doi <DOI> --sourcePath <Source Path>

Required:
--doi            Specify DOI to process
--sourcePath     Specify source path
--wasabi         Copy files to Wasabi bucket
--backup         Copy files to s3 bucket

Available Option:
--apply          Escape dry run mode

Example usages:
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi --apply
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --backup
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --backup --apply
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi --backup
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi --backup --apply
[lily@ip-10-99-0-212 ~]$ /usr/local/bin/transfer --doi 102481 --sourcePath /share/dropbox/user101/
Error: please specify --wasabi or --backup or both
Usage: /usr/local/bin/transfer --doi <DOI> --sourcePath <Source Path>

Required:
--doi            Specify DOI to process
--sourcePath     Specify source path
--wasabi         Copy files to Wasabi bucket
--backup         Copy files to s3 bucket

Available Option:
--apply          Escape dry run mode

Example usages:
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi --apply
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --backup
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --backup --apply
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi --backup
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi --backup --apply
[lily@ip-10-99-0-212 ~]$ /usr/local/bin/transfer --doi 102481 --sourcePath /share/dropbox/user101/ --wasabi
[lily@ip-10-99-0-212 ~]$ ls /var/log/gigadb/
backup_20240805_041137.log
[lily@ip-10-99-0-212 ~]$ /usr/local/bin/transfer --doi 102481 --sourcePath /share/dropbox/user101/ --wasabi --backup
[lily@ip-10-99-0-212 ~]$ ls /var/log/gigadb/
backup_20240805_041137.log  backup_20240805_041332.log
[lily@ip-10-99-0-212 ~]$ cat /var/log/gigadb/backup_20240805_041332.log 
2024/08/05 04:13:32 INFO  : Start copying files from staging to Wasabi
2024/08/05 04:13:33 NOTICE: readme_102480.txt: Skipped copy as --dry-run is set (size 3.127Ki)
2024/08/05 04:13:33 NOTICE: analysis_data/Tree_file.txt: Skipped copy as --dry-run is set (size 359)
2024/08/05 04:13:33 NOTICE: 
Transferred:        3.478 KiB / 3.478 KiB, 100%, 0 B/s, ETA -
Transferred:            2 / 2, 100%
Elapsed time:         0.3s

2024/08/05 04:13:33 INFO  : Executed: rclone copy --s3-no-check-bucket /share/dropbox/user101/ wasabi:gigadb-datasets/staging/pub/10.5524/102001_103000/102481 --dry-run --log-file /var/log/gigadb/backup_20240805_041332.log --log-level INFO --stats-log-level DEBUG >> /var/log/gigadb/backup_20240805_041332.log
2024/08/05 04:13:33 INFO  : Successfully copied files to Wasabi bucket for DOI: 102481
2024/08/05 04:13:33 INFO  : Start copying files from staging to s3
2024/08/05 04:13:33 NOTICE: readme_102480.txt: Skipped copy as --dry-run is set (size 3.127Ki)
2024/08/05 04:13:33 NOTICE: analysis_data/Tree_file.txt: Skipped copy as --dry-run is set (size 359)
2024/08/05 04:13:33 NOTICE: 
Transferred:        3.478 KiB / 3.478 KiB, 100%, 0 B/s, ETA -
Transferred:            2 / 2, 100%
Elapsed time:         0.2s

2024/08/05 04:13:33 INFO  : Executed: rclone copy --s3-no-check-bucket /share/dropbox/user101/ gigadb-datasetfiles:gigadb-datasetfiles-backup/staging/pub/10.5524/102001_103000/102481 --dry-run --log-file /var/log/gigadb/backup_20240805_041332.log --log-level INFO --stats-log-level DEBUG >> /var/log/gigadb/backup_20240805_041332.log
2024/08/05 04:13:33 INFO  : Successfully copied files to s3 bucket for DOI: 102481
[lily@ip-10-99-0-212 ~]$ 

```

### As a centos user in live
```
% ssh -i path/to/live/pem centos@$live-bastion-ip
Activate the web console with: systemctl enable --now cockpit.socket

Last login: Mon Aug  5 04:48:09 2024 from 54.180.33.208
[centos@ip-10-99-0-253 ~]$ df -hT
Filesystem     Type      Size  Used Avail Use% Mounted on
devtmpfs       devtmpfs  339M     0  339M   0% /dev
tmpfs          tmpfs     372M     0  372M   0% /dev/shm
tmpfs          tmpfs     372M  5.3M  367M   2% /run
tmpfs          tmpfs     372M     0  372M   0% /sys/fs/cgroup
/dev/nvme0n1p1 xfs        30G  2.5G   28G   9% /
tmpfs          tmpfs      75M     0   75M   0% /run/user/1000
127.0.0.1:/    nfs4      8.0E     0  8.0E   0% /share/dropbox
127.0.0.1:/    nfs4      8.0E     0  8.0E   0% /share/config
[centos@ip-10-99-0-253 ~]$ cat files-env 
GIGADB_ENV=live
WASABI_DATASETFILES_DIR=wasabi:gigadb-datasets/live/pub/10.5524
S3_DATASETFILES_DIR=gigadb-datasetfiles:gigadb-datasetfiles-backup/live/pub/10.5524
[centos@ip-10-99-0-253 ~]$ ls /share/dropbox/user101/
analysis_data  readme_102480.txt
[centos@ip-10-99-0-253 ~]$ ls /share/dropbox/user101/analysis_data/
Tree_file.txt
[centos@ip-10-99-0-253 ~]$ /usr/local/bin/transfer
Usage: /usr/local/bin/transfer --doi <DOI> --sourcePath <Source Path>

Required:
--doi            Specify DOI to process
--sourcePath     Specify source path
--wasabi         Copy files to Wasabi bucket
--backup         Copy files to s3 bucket

Available Option:
--apply          Escape dry run mode

Example usages:
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi --apply
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --backup
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --backup --apply
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi --backup
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi --backup --apply
[centos@ip-10-99-0-253 ~]$ /usr/local/bin/transfer --doi 102480 --sourcePath /share/dropbox/user101/
Error: please specify --wasabi or --backup or both
Usage: /usr/local/bin/transfer --doi <DOI> --sourcePath <Source Path>

Required:
--doi            Specify DOI to process
--sourcePath     Specify source path
--wasabi         Copy files to Wasabi bucket
--backup         Copy files to s3 bucket

Available Option:
--apply          Escape dry run mode

Example usages:
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi --apply
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --backup
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --backup --apply
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi --backup
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi --backup --apply
[centos@ip-10-99-0-253 ~]$ ls /var/log/gigadb/
[centos@ip-10-99-0-253 ~]$ /usr/local/bin/transfer --doi 102480 --sourcePath /share/dropbox/user101/ --wasabi --backup
[centos@ip-10-99-0-253 ~]$ ls /var/log/gigadb/
backup_20240805_054904.log
[centos@ip-10-99-0-253 ~]$ cat /var/log/gigadb/backup_20240805_054904.log 
2024/08/05 05:49:04 INFO  : Start copying files from live to Wasabi
2024/08/05 05:49:05 NOTICE: readme_102480.txt: Skipped copy as --dry-run is set (size 3.127Ki)
2024/08/05 05:49:05 NOTICE: analysis_data/Tree_file.txt: Skipped update modification time as --dry-run is set (size 359)
2024/08/05 05:49:05 NOTICE: 
Transferred:        3.127 KiB / 3.127 KiB, 100%, 0 B/s, ETA -
Checks:                 2 / 2, 100%
Transferred:            1 / 1, 100%
Elapsed time:         1.5s

2024/08/05 05:49:05 INFO  : Executed: rclone copy --s3-no-check-bucket /share/dropbox/user101/ wasabi:gigadb-datasets/live/pub/10.5524/102001_103000/102480 --dry-run --log-file /var/log/gigadb/backup_20240805_054904.log --log-level INFO --stats-log-level DEBUG >> /var/log/gigadb/backup_20240805_054904.log
2024/08/05 05:49:05 INFO  : Successfully copied files to Wasabi bucket for DOI: 102480
2024/08/05 05:49:05 INFO  : Start copying files from live to s3
2024/08/05 05:49:06 NOTICE: readme_102480.txt: Skipped copy as --dry-run is set (size 3.127Ki)
2024/08/05 05:49:06 NOTICE: analysis_data/Tree_file.txt: Skipped copy as --dry-run is set (size 359)
2024/08/05 05:49:06 NOTICE: 
Transferred:        3.478 KiB / 3.478 KiB, 100%, 0 B/s, ETA -
Transferred:            2 / 2, 100%
Elapsed time:         0.2s

2024/08/05 05:49:06 INFO  : Executed: rclone copy --s3-no-check-bucket /share/dropbox/user101/ gigadb-datasetfiles:gigadb-datasetfiles-backup/live/pub/10.5524/102001_103000/102480 --dry-run --log-file /var/log/gigadb/backup_20240805_054904.log --log-level INFO --stats-log-level DEBUG >> /var/log/gigadb/backup_20240805_054904.log
2024/08/05 05:49:06 INFO  : Successfully copied files to s3 bucket for DOI: 102480
[centos@ip-10-99-0-253 ~]$ 
```

### As a user lily in live
```
% cd ops/infrastructure/envs/staging
% ansible-playbook -i ../../inventories users_playbook.yml -e "newuser=lily" -e "credentials_csv_path=~/Downloads/credentials.csv" -e "gigadb_env=live"
% chmod 500 output/privkeys-$bastion-ip/lily
% ssh -i utput/privkeys-54.180.33.208/lily lily@$bastion-ip
[lily@ip-10-99-0-253 ~]$ ls
files-env  uploadDir
[lily@ip-10-99-0-253 ~]$ df -hT
Filesystem     Type      Size  Used Avail Use% Mounted on
devtmpfs       devtmpfs  339M     0  339M   0% /dev
tmpfs          tmpfs     372M     0  372M   0% /dev/shm
tmpfs          tmpfs     372M  9.9M  362M   3% /run
tmpfs          tmpfs     372M     0  372M   0% /sys/fs/cgroup
/dev/nvme0n1p1 xfs        30G  2.5G   28G   9% /
tmpfs          tmpfs      75M     0   75M   0% /run/user/1000
127.0.0.1:/    nfs4      8.0E     0  8.0E   0% /share/dropbox
127.0.0.1:/    nfs4      8.0E     0  8.0E   0% /share/config
tmpfs          tmpfs      75M     0   75M   0% /run/user/1001
[lily@ip-10-99-0-253 ~]$ cat files-env 
GIGADB_ENV=live
WASABI_DATASETFILES_DIR=wasabi:gigadb-datasets/live/pub/10.5524
S3_DATASETFILES_DIR=gigadb-datasetfiles:gigadb-datasetfiles-backup/live/pub/10.5524
[lily@ip-10-99-0-253 ~]$ ls /share/dropbox/user101/
analysis_data  readme_102480.txt
[lily@ip-10-99-0-253 ~]$ ls /share/dropbox/user101/analysis_data/
Tree_file.txt
[lily@ip-10-99-0-253 ~]$ ls /var/log/gigadb/
backup_20240805_054904.log
[lily@ip-10-99-0-253 ~]$ /usr/local/bin/transfer
Usage: /usr/local/bin/transfer --doi <DOI> --sourcePath <Source Path>

Required:
--doi            Specify DOI to process
--sourcePath     Specify source path
--wasabi         Copy files to Wasabi bucket
--backup         Copy files to s3 bucket

Available Option:
--apply          Escape dry run mode

Example usages:
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi --apply
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --backup
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --backup --apply
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi --backup
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi --backup --apply
[lily@ip-10-99-0-253 ~]$ /usr/local/bin/transfer --doi 102480 --sourcePath /share/dropbox/user101/
Error: please specify --wasabi or --backup or both
Usage: /usr/local/bin/transfer --doi <DOI> --sourcePath <Source Path>

Required:
--doi            Specify DOI to process
--sourcePath     Specify source path
--wasabi         Copy files to Wasabi bucket
--backup         Copy files to s3 bucket

Available Option:
--apply          Escape dry run mode

Example usages:
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi --apply
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --backup
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --backup --apply
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi --backup
/usr/local/bin/transfer --doi 100148 --sourcePath /share/dropbox/user101 --wasabi --backup --apply
[lily@ip-10-99-0-253 ~]$ ls /var/log/gigadb/
backup_20240805_054904.log  backup_20240805_061241.log
[lily@ip-10-99-0-253 ~]$ cat /var/log/gigadb/backup_20240805_061241.log 
2024/08/05 06:12:41 INFO  : Start copying files from live to Wasabi
2024/08/05 06:12:43 NOTICE: readme_102480.txt: Skipped copy as --dry-run is set (size 3.127Ki)
2024/08/05 06:12:43 NOTICE: analysis_data/Tree_file.txt: Skipped update modification time as --dry-run is set (size 359)
2024/08/05 06:12:43 NOTICE: 
Transferred:        3.127 KiB / 3.127 KiB, 100%, 0 B/s, ETA -
Checks:                 2 / 2, 100%
Transferred:            1 / 1, 100%
Elapsed time:         1.4s

2024/08/05 06:12:43 INFO  : Executed: rclone copy --s3-no-check-bucket /share/dropbox/user101/ wasabi:gigadb-datasets/live/pub/10.5524/102001_103000/102480 --dry-run --log-file /var/log/gigadb/backup_20240805_061241.log --log-level INFO --stats-log-level DEBUG >> /var/log/gigadb/backup_20240805_061241.log
2024/08/05 06:12:43 INFO  : Successfully copied files to Wasabi bucket for DOI: 102480
2024/08/05 06:12:43 INFO  : Start copying files from live to s3
2024/08/05 06:12:43 NOTICE: readme_102480.txt: Skipped copy as --dry-run is set (size 3.127Ki)
2024/08/05 06:12:43 NOTICE: analysis_data/Tree_file.txt: Skipped copy as --dry-run is set (size 359)
2024/08/05 06:12:43 NOTICE: 
Transferred:        3.478 KiB / 3.478 KiB, 100%, 0 B/s, ETA -
Transferred:            2 / 2, 100%
Elapsed time:         0.2s

2024/08/05 06:12:43 INFO  : Executed: rclone copy --s3-no-check-bucket /share/dropbox/user101/ gigadb-datasetfiles:gigadb-datasetfiles-backup/live/pub/10.5524/102001_103000/102480 --dry-run --log-file /var/log/gigadb/backup_20240805_061241.log --log-level INFO --stats-log-level DEBUG >> /var/log/gigadb/backup_20240805_061241.log
2024/08/05 06:12:43 INFO  : Successfully copied files to s3 bucket for DOI: 102480
```

## How have functionalities been implemented?

Describe how the new functionalities have been implemented by the
changed code at a high level

Created a wrapper script `efs_backup.sh` which will call rclone to copy files from a source to a destinated wasabi bucket and also aws s3 bucket in glacier tier. By default, it will copy files to destinationation in dry-run mode, unless `--apply` flag is supplied. And also, for the destination, it requires at least `--wasabi` or `--backup` is supplied, otherwise, the script will throw `Error: please specify --wasabi or --backup or both`. 

For the destinaton bucket path, it would be determinated by the environment (eg. dev, staging, live) of the script being executed. And the bucket path variables need to be created by each developer into their gitlab repo project page.
And all the information will be saved in the `files-env` file, the script will get the values on the fly.

The `webapp_playbook` and `bastion_playbook` has also been updated to mount aws efs accesspoints, the `ansible_init.sh` will save the efs id and the accesspoint ids into the ansible.properties which then will be populated by ansible to execute the mounting actions. The [aws-efs-utils](https://github.com/aws/efs-utils?tab=readme-ov-file) has been pre-compiled and the rpm has been saved to the wasabi bucket `infra-resources`, the current version is `amazon-efs-utils-2.0.4-1.el8.x86_64.rpm`. This completed https://github.com/gigascience/gigadb-website/issues/1771


## Any issues with implementation?

None.

## Any changes to automated tests?

None.

## Any changes to documentation?

None.

## Any technical debt repayment?

Removed the immature `mount-efs.sh` from the terraform side, as there are ansible tasks for the mounting.

For the user rclone config `ops/infrastructure/roles/bastion-users/tasks/rclone-user.conf.j2`, the name suffix of `wasabi` destination has been removed, so user just need to use `wasabi:$bucketPath` to specify the end point, instead of `wasabi_$user:$bucketpath`, which fixed https://github.com/gigascience/gigadb-website/issues/1861

## Any improvements to CI/CD pipeline?

None.

kencho51 committed:
doc: create policy for bucket gigadb-datasetfiles-backup

kencho51 committed:
refactor: rename the policy

kencho51 committed:
refactor: remove renamed file

kencho51 committed:
doc: update the sid name

kencho51 committed:
refactor: update config for accessing datasetfiles backup bucket

kencho51 committed:
refactor: get the secrets for the datasetfiles backup bucket

kencho51 committed:
refactor: untrack log dir

kencho51 committed:
feat: wrapper script for upload files to s3 and wasabi

kencho51 committed:
refactor: make datasetfiles backup bucket credential available in dev

kencho51 committed:
feat: make script available in the bastion server

kencho51 committed:
refactor: pick the right destination paths in different envs

kencho51 committed:
refactor: add in remote name

kencho51 committed:
refactor: remove user name from the remote name section

kencho51 committed:
refactor: make the tests check for the bucket

kencho51 committed:
refactor: reset the bucket state as a teardown

kencho51 committed:
feat: make efs ids available for ansible

kencho51 committed:
refactor: use efs dns name for mounting

kencho51 committed:
feat: create task for mounting access points

kencho51 committed:
refactor: remove data script for mounting in terraform

kencho51 committed:
feat: make webapp server has config access point

kencho51 committed:
refactor: update the host name

kencho51 committed:
refactor: update the dir for the log output

kencho51 committed:
refactor: make the script use secrets and env

kencho51 committed:
feat: make bucket info available in production envs

kencho51 committed:
refactor: use exact path for getting files-env

kencho51 committed:
refactor: create /var/log/gigadb for storing the log outputs

kencho51 committed:
refactor: make wasabi or backup flags compulsory

kencho51 committed:
doc: update change log

kencho51 committed:
refactor: make the log accessibly to any user

kencho51 committed:
refactor: make the files-env available to the user

kencho51 committed:
refactor: make the script executable

kencho51 committed:
doc: update the change log after rebase

rija reviewed (commented):
Hi @kencho51,
**chore**: could you rebase your PR onto HEAD of develop to incorporate the pipeline doc PR that was merged recently, as my pipeline of your branch is failing and the recently merged PR has a fix for it.

rija reviewed (commented):
Hi @kencho51,

**question**: how were created the AWS access keys used as value for `gigadb_dataset_metadata_aws_access_key_id` and `gigadb_datasets_metadata_aws_secret_access_key` ?
**issue**: the dependency between the projects `wasabi-migration` and `files-metadata-console` is not a good idea: it is not necessary, it complicate deployment (you cannot deploy one without the other) and testing, and it will confuse developers who will need to work on this in the future. (only `gigadb/app/configurator` is ok to be depended upon)
**issue**: the file-metadata console project's purpose is any action that's related to the Gigadb database's files table and associated tables. Backup and publishing of actual files don't really fit in there.
**issue**: `efs_backup` is misleading, as it's only half of its job. Something like `transfer-files` or simply `transfer` is more representative and easier to remember.
**suggestion**: Create new project for file transfers operations at `gigadb/app/tools/transfer-files` and name the command `transfer`. The config file, the bats tests and the command needed by the project should all be in this project directory.
**issue**: the bats tests are all failing locally with access denied error thrown by rclone for both S3 and Wasabi. 
My wasabi access keys have been added to the generated rclone.conf, so I don't know why the errors.
However, the `gigadb_dataset_metadata_aws_access_key_id` and `gigadb_datasets_metadata_aws_secret_access_key` haven't been added to the rclone.conf by the configuration script.
**praise**: all tests are passing on CI pipeline

kencho51 commented:
Hi @rija,

> **issue**: the bats tests are all failing locally with access denied error thrown by rclone for both S3 and Wasabi.
> My wasabi access keys have been added to the generated rclone.conf, so I don't know why the errors.
> However, the `gigadb_dataset_metadata_aws_access_key_id` and `gigadb_datasets_metadata_aws_secret_access_key` haven't been added to the rclone.conf by the configuration script.

These two new variables are added into the `cnhk-infra` variable page, they will be added to the `gigadb/app/tools/wasabi-migration/config/rclone.conf` after executing the `docker-compose run --rm config` in `gigadb/app/tools/wasabi-migration`.

So, have you removed the existing `.secrets` and executed the `docker-compose run --rm config` cmd? The details actually can refer to the `How to test` `In dev` section in this PR description.

kencho51 committed:
refactor: create transfer-files dir

kencho51 committed:
refactor: revert changes to the existing files in dev

kencho51 committed:
refactor: update playbooks with the updated tool name

kencho51 committed:
refactor: remove extra config

kencho51 committed:
refactor: update the script source path to copy

kencho51 committed:
doc: create readme for the tool

kencho51 committed:
refactor: update the log file name

kencho51 committed:
refactor: remove extra test files

kencho51 committed:
doc: update log file name

kencho51 commented:
Hi @rija 

> **suggestion**: Create new project for file transfers operations at `gigadb/app/tools/transfer-files` and name the command `transfer`. The config file, the bats tests and the command needed by the project should all be in this project directory.

A new project `gigadb/app/tools/transfer-files` is created to hold the new tool.

Please take a look at this PR again.

rija reviewed (commented):
Hi @kencho51,

**praise**: great to see the tool being in self-contained project with good naming
**issue** (non-blocking): Your instructions were missing a step to ensure WASABI_ACCESS_KEY_ID and WASABI_SECRET_ACCESS_KEY exist in the Gitlab project's variables
**praise**: all the bats tests are now passing after the aforementioned variables were added 
**issue**: there is big block of duplicated code in `generate_config.sh`, the block that download and reassemble gitlab variables is needed in all our projects, so there is a library for that in `gigadb/app/configurator`. You can look in `gigadb/app/tools/files-metadata-console/configure` for an example of use
**issue**: the change to the terraform output is a dangerous change:
```
Changes to Outputs:
  - efs_filesystem_access_points         = {
      - configuration_area = "fsap-0f97566312267b7bc"
      - dropbox_area       = "fsap-04c75916a421bffe7"
    } -> null
  + efs_filesystem_configuration_area_id = "fsap-0f97566312267b7bc"
  + efs_filesystem_dropbox_area_id       = "fsap-04c75916a421bffe7"
  ```

It means, that on live production we would have to run terraform and based on the plan I have just run from the current v4.3.3 release, a certain number of resources will need to be destroyed and re-created.

It'd be safer doing that when the blue/green infrastructure provisioning is available. 

Why don't you use the `efs_filesystem_access_points` block? It already has the information needed for mounting the EFS.

kencho51 committed:
refactor: move all template files to config-source dir

kencho51 committed:
refacfor: remove generate config files

kencho51 committed:
refactor: create configure script to generate configs

kencho51 committed:
refactor: make dotfiles get all necessary variables

kencho51 committed:
refactor: update path for the files-env file

kencho51 committed:
doc: update the step to generate config

kencho51 commented:
Hi Rija,

> **issue**: the change to the terraform output is a dangerous change:

I noticed that I have updated the tf code which would bring some changes in the existing infrastructure, but just the output part. 
First, the reason for the change in tf code because I want to pass the ids of the aws efs system to the ansible playbooks by this code block in the `ansible_init.sh`
```
# Retrieve efs ids for mounting
efs_filesystem_dns_name=$(terraform output efs_filesystem_dns_name | sed 's/"//g')
configuration_area_id=$(terraform output efs_filesystem_configuration_area_id | sed 's/"//g')
dropbox_area_id=$(terraform output efs_filesystem_dropbox_area_id | sed 's/"//g')
```
These ids will then be added into the `ansible.properties` which can be populated by the playbooks for the mounting.
Second, when I switched to this PR and tried:
```
(upload-dataset-from-efs-to-wasabi-and-glacier$=) % ../../../scripts/tf_init.sh --project gigascience/forks/kencho-gigadb-website --env staging
(upload-dataset-from-efs-to-wasabi-and-glacier$=) % terraform plan
...
Plan: 0 to add, 4 to change, 0 to destroy.

Changes to Outputs:
  - efs_filesystem_access_points         = {
      - configuration_area = "fsap-05f325ad28cbf4264"
      - dropbox_area       = "fsap-061df6699cf4f7312"
    } -> null
  + efs_filesystem_configuration_area_id = "fsap-05f325ad28cbf4264"
  + efs_filesystem_dropbox_area_id       = "fsap-061df6699cf4f7312"
```

The fsap ids keeps unchanged, so no new efs system has been created, only the output format of the ids is changed.

Otherwise, I could revert the changes in the tf code and get the efs ids in the `ansible_init.sh` as below:
```
configuration_area_id=$(terraform output -json efs_filesystem_access_points | jq -r '.configuration_area')
```

As discussed in the meeting, I have reverted the changes made in the tf output part.

Please look again.

kencho51 committed:
refactor: revert the changes made in the tf output

rija reviewed (commented):
Hi @kencho51,

**praise**: Good to see that there is no need to run terraform for this PR, making it possible for safe deploy on the live production environment 
**praise**: Good to see the new S3 IAM policies documented
**praise**: Good to see a README.md in the tool's directory

**issue**: Ansible third party collection posix has to be run manually
**suggestion**: We have a file `ops/infrastructure/requirements.yml` where we put all Ansible plugins, so they can be installed/updated together as documented in our SOPs, so there should be a new entry in that file for the posix collection

**issue** (non-blocking): the command `transfer` should output the path to the log file after it has run
**suggestion**: for end user convenience, the path to the log should be printed on screen when the command completes, so the user can copy/paste it

**issue**: There should not be a different log file for each run of the transfer command.
**suggestion**: Current implementation make it hard to locate and review the right log file as we need to remember the time we started a particular session, and if  the transfer took a long time, and there already a lot of log files, the user may have forgotten the start time and will get confused. And by the time we found the right file, we still need to copy paste that specific file name to view it. 
Additionally, it is not a standard way for UNIX/Linux commands which means that we cannot easily use other log analysis/houskeeping commands to work with our logs files (In particular, I'm thinking of `logrotate` to automatically housekeep the logs of our commands).
There should just be one log file that keep being appended to: `transfer.log`

**issue**: In non-centos user's `.aws/credentials`, the Wasabi credentials are stored under `[default]`
**suggestion**: It is not clear that the pair of key in `.aws/credentials` is a Wasabi key, and it may be confused for an AWS one, which can be problematic if we want the non-centos user to access other AWS services in the future. I suggest to replace `[default]` with `[wasabi]`

**issue**: In non-centos user's `.config/rclone/rclone.conf`, the value of access keys and secrets keys are present
**suggestion**: The issue here is that we should try to not disseminate sensitive information in multiple file. `.aws/credentials` is the standard location for all AWS-compatible services, so the access keys to access the S3 bucket should be located there under an explicit profile name (e.g: `[gigadb_aws]`) 

**issue**: Large blocks of code snippets that mix command example and instructions/explanations are not user-friendly, and I can see it in the testing instructions for this PR and in the `README.md`  for the `transfer-files` project
**suggestion**: Each single command example should be surrounded by triple quotes (so they can be copy/pasted easily without fiddling). If output needs to be shown, it should be in a separate code block. Any instructions/explanations is not code example so it should be in the main document, not in a code block, otherwise they are hard to parse and can be easily missed.

**praise**: functionality works as described in the testing instructions
**praise**: the mounting of the EFS share is now automated
**praise**: the transfer command usage looks neat and exhaustive

kencho51 committed:
refactor: make sure config dir exists

kencho51 committed:
refactor: make the script output log path

kencho51 committed:
refactor: install posix as a collection

kencho51 committed:
refactor: remove secrets from the rclone config

kencho51 committed:
refactor: make the script to use s3 profile for using different credentials

kencho51 committed:
refactor: reformate the credential template

kencho51 committed:
refactor: update the production rclone user config to use env auth

kencho51 commented:
Hi @rija,

> **issue**: Ansible third party collection posix has to be run manually

Fixed in https://github.com/gigascience/gigadb-website/pull/1977/commits/818513db137358ced26ace4718f5ea3f625c05f1

>**issue (non-blocking)**: the command transfer should output the path to the log file after it has run
>**issue**: There should not be a different log file for each run of the transfer command.

Fixed in https://github.com/gigascience/gigadb-website/pull/1977/commits/899e6ed4634892e9e71a250b722d403f6508ed5d, as below screen output in my staging
```
[lily@ip-10-99-0-28 ~]$ /usr/local/bin/transfer --doi 102481 --sourcePath /share/dropbox/user101/ --wasabi --backup
More details about copying files to Wasabi bucket, please refer to: /var/log/gigadb/transfer.log
More details about copying files to s3 bucket, please refer to: /var/log/gigadb/transfer.log
[lily@ip-10-99-0-28 ~]$ less /var/log/gigadb/transfer.log | tail
2024/09/13 08:19:42 NOTICE: readme_102480.txt: Skipped update modification time as --dry-run is set (size 3.127Ki)
2024/09/13 08:19:42 NOTICE: analysis_data/Tree_file.txt: Skipped update modification time as --dry-run is set (size 359)
2024/09/13 08:19:42 INFO  : There was nothing to transfer
2024/09/13 08:19:42 NOTICE: 
Transferred:              0 B / 0 B, -, 0 B/s, ETA -
Checks:                 2 / 2, 100%
Elapsed time:         0.5s

2024/09/13 08:19:42 INFO  : Executed: rclone copy --s3-no-check-bucket --s3-profile aws-transfer /share/dropbox/user101/ gigadb-datasetfiles:gigadb-datasetfiles-backup/staging/pub/10.5524/102001_103000/102481 --dry-run --log-file /var/log/gigadb/transfer.log --log-level INFO --stats-log-level DEBUG >> /var/log/gigadb/transfer.log
2024/09/13 08:19:42 INFO  : Successfully copied files to s3 bucket for DOI: 102481
```


>**issue**: In non-centos user's .aws/credentials, the Wasabi credentials are stored under [default]
>**issue**: In non-centos user's .config/rclone/rclone.conf, the value of access keys and secrets keys are present

Fixed in https://github.com/gigascience/gigadb-website/pull/1977/commits/c1169aa38bfe2e80383099f01e19d82d9c1f9ca1 by introducing `s3-profile` parameters into the rclone cmd
And the profile in the `.aws/credentials` is named by `serviceProvider-application`, eg. `aws-transfer` and `wasabi-transfer`.

For detail usage of the `transfer.sh` and the rclone cmd involved, please refer to the `tests/bats/transfer.bats.

Please have a look at this PR again, thanks.

kencho51 committed:
refactor: make s3 profile parameter available for the centos user

kencho51 committed:
refactor: make the aws credentials available for centos user

kencho51 committed:
doc: update readme to include s3 profile instruction

kencho51 committed:
refactor: fix the csp to allow execution of external script and style from jsdelivr

kencho51 commented:
Hi @Rija,

The latest [commit](https://github.com/gigascience/gigadb-website/pull/1977/commits/38645ec93dd0b2b47d611ec02d2a94c8ff59700c) is trying to fix the CSP problem in ticket https://github.com/gigascience/gigadb-website/issues/1944

```
Refused to load the script 'https://cdn.jsdelivr.net/npm/ol@v8.1.0/dist/ol.js' because it violates the following Content Security Policy directive: "script-src 'self' 'unsafe-inline' 'unsafe-eval' https://cdn.matomo.cloud/gigadb.matomo.cloud/matomo.js *.hypothes.is *.datatables.net *.cloudflare.com *.google-analytics.com https://www.rosaceae.org https://www.protocols.io https://hypothes.is https://codeocean.com https://tumormap.ucsc.edu https://openlayers.org/en/v4.6.5/build/ol.js". Note that 'script-src-elem' was not explicitly set, so 'script-src' is used as a fallback.

Refused to load the stylesheet 'https://cdn.jsdelivr.net/npm/ol@v8.1.0/ol.css' because it violates the following Content Security Policy directive: "default-src 'self' 'unsafe-inline' 'unsafe-eval' http://gigadb.org http://penguin.genomics.cn https://www.rosaceae.org *.protocols.io https://sketchfab.com https://codeocean.com *.hypothes.is *.datatables.net *.cloudflare.com". Note that 'style-src-elem' was not explicitly set, so 'default-src' is used as a fallback.

Refused to connect to 'https://gigadb.matomo.cloud/matomo.php?action_name=GigaDB%20-%20Map%20Browse&idsite=1&rec=1&r=794844&h=14&m=51&s=23&url=https%3A%2F%2Fgigadb.org%2Fsite%2Fmapbrowse&_id=cb7a33a88f8d43c9&_idn=0&send_image=0&_refts=0&pv_id=5ObhFB&fa_pv=1&fa_fp[0][fa_vid]=NvRO5x&fa_fp[0][fa_fv]=1&pf_net=0&pf_srv=227&pf_tfr=1&uadata=%7B%22fullVersionList%22%3A%5B%7B%22brand%22%3A%22Chromium%22%2C%22version%22%3A%22128.0.6613.138%22%7D%2C%7B%22brand%22%3A%22Not%3BA%3DBrand%22%2C%22version%22%3A%2224.0.0.0%22%7D%2C%7B%22brand%22%3A%22Google%20Chrome%22%2C%22version%22%3A%22128.0.6613.138%22%7D%5D%2C%22mobile%22%3Afalse%2C%22model%22%3A%22%22%2C%22platform%22%3A%22macOS%22%2C%22platformVersion%22%3A%2213.4.1%22%7D&pdf=1&qt=0&realp=0&wma=0&fla=0&java=0&ag=0&cookie=1&res=2056x1329' because it violates the following Content Security Policy directive: "default-src 'self' 'unsafe-inline' 'unsafe-eval' http://gigadb.org http://penguin.genomics.cn https://www.rosaceae.org *.protocols.io https://sketchfab.com https://codeocean.com *.hypothes.is *.datatables.net *.cloudflare.com". Note that 'connect-src' was not explicitly set, so 'default-src' is used as a fallback.
```

After deployment to production, the above errors should be cleared from the browsers.

kencho51 committed:
doc: update change log

rija reviewed (approved):
Hi @kencho51,

**praise**: requested changes implemented
**praise**: mapbrowse now appears

happy to approve

pli888 commented:
@kencho51 For the record, running `./configure` produces an error:
```
$ ./configure
An .env file is present
Sourcing .env
Retrieving variables from https://gitlab.com/api/v4/groups/gigascience/variables?per_page=100
Retrieving variables from https://gitlab.com/api/v4/groups/3501869/variables
Retrieving variables from https://gitlab.com/api/v4/projects/gigascience%2Fforks%2Fpli888-gigadb-website/variables
Retrieving variables from https://gitlab.com/api/v4/projects/gigascience%2Fcnhk-infra/variables
Sourcing secrets
Current environment: dev
../../configurator/src/dotfiles.sh: line 93: $3: unbound variable
```

However, we discussed at today's VOH that `config-sources/rclone.conf.dist` is not required because no variables whose values need to be interpolated are required to create a `rclone.conf`  file.

pli888 commented on gigadb/app/tools/transfer-files/README.md:
````diff
@@ -0,0 +1,513 @@
+### In dev
+```
+# update the rclone config file
+% cd gigadb/app/tools/transfer-files
````
**nitpick**: With all the commands mixed in with command-line output within one single code block, it's easy to miss a step that I need to perform.

pli888 commented on gigadb/app/tools/transfer-files/config-sources/rclone.conf.dist:
```diff
@@ -0,0 +1,18 @@
+[wasabi]
```
**question**: We do not need this template file because there are no variable values to interpolate in order to create `rclone.conf`?

kencho51 replied:
**answer:** The dist file and its related steps have been removed in https://github.com/gigascience/gigadb-website/pull/1977/commits/f0328c5a596b1b6e663e4a920159b00972b018d0

pli888 commented on gigadb/app/tools/transfer-files/configure:
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env bash
```
question: We do not need this configure script because we do not need it to create a rclone.conf file?

kencho51 replied:
**answer:** We still need it to generate `.env` and `.secrets` to execute the `transfer.sh` locally.

pli888 reviewed (changes requested):
**praise**: The automatic connection of EFS to bastion server functionality is working on my staging environment.
**praise**: `transfer.sh` seems to be working for me.
**question**: Sample map is visible but it displays no samples. Is this expected due to another problem, e.g. sample co-ordinates not being correctly formatted?
**issue**: Need to remove use of template file for creating rclone configuration file.
**issue (non-blocking)**: Dev testing instructions in PR report says we need $gigadb_dataset_metadata_aws_access_key_id and $gigadb_datasets_metadata_aws_secret_access_key in the `[aws-transfer]` section in `~/.aws/credentials`. We actually need to use `$gigadb_datasetfiles_aws_access_key_id` and `$gigadb_datasetfiles_aws_secret_access_key`.
**info**: $gigadb_dataset_metadata_aws_access_key_id and $gigadb_datasets_metadata_aws_secret_access_key will need to be deleted after PR #2018 bastion tools for curators has been merged and deployed onto live as we don't need this bucket anymore.
**issue (non-blocking)**: Two tests are failing on my CI pipeline: `FileUrlsUpdaterTest` and `ReadmeGeneratorTest` for the same reason: The repository 'http://apt.postgresql.org/pub/repos/apt buster-pgdg Release' does not have a Release file.
**suggestion**: https://www.debian.org/releases/buster/ says Debian 10 buster has been superceded by Debian 11 bullseye and buster has no security updates since Jun 2022. http://apt.postgresql.org/pub/repos/apt does not seem to have packages for buster so this is probably cause of the error and we need to move to bullseye distro for dockerfile base image. Create a separate Github ticket for this work?
**issue**: On my GigaDB home page on my staging server, I see a strange box element at the bottom of the web page. Do you see this too?
<img width="1470" alt="gigadb_problem" src="https://github.com/user-attachments/assets/76d65c49-0a78-42bc-9120-60c18f3f9f16">

kencho51 commented:
Hi @pli888 ,

> **question**: Sample map is visible but it displays no samples. Is this expected due to another problem, e.g. sample co-ordinates not being correctly formatted?

Yes, although the CSP has been updated, the map will still not showing any sample, details at https://github.com/gigascience/gigadb-website/issues/1944

>**issue**: On my GigaDB home page on my staging server, I see a strange box element at the bottom of the web page. Do you see this too?

I can see this `hypothesis` box on my staging too, will look into it. I was suspecting that the updated CSP is the cause of it.

kencho51 committed:
refactor: Remove extra dist file

kencho51 committed:
refactor: update the aws key pairs name

kencho51 committed:
Revert "refactor: fix the csp to allow execution of external script and style from jsdelivr"

This reverts commit 91e091f9ac75de8b704b3ef78ef0d9d8e66c7680.

kencho51 committed:
Revert "doc: update change log"

This reverts commit 3dfce0cc9e088f33cc52df18e615f1839b77d96c.

kencho51 commented:
Hi @pli888 @rija,

I can confirm that the hypothesis box would appear normally as in current https://gigadb.org after revert the changes in the CSP config.

Since it requires time to tweak CSP for the correct styling, so I have reverted the related CSP commits from this PR and will try to fix it in PR https://github.com/gigascience/gigadb-website/pull/2010.

Please have a look again.

kencho51 committed:
doc: remove extra entry

kencho51 committed:
refactor: update mint doi error message

kencho51 committed:
refactor: update CSP to display hypotheis and mapbrowse properly

kencho51 committed:
doc: update change log for CSP

kencho51 commented:
Hi @pli888 @rija,

After playing with the CSP, this latest commit https://github.com/gigascience/gigadb-website/pull/1977/commits/e149271f76a5aecab33e09a0d6312e40db746cc1: 
 - allows script execution from `https://cdn.jsdelivr.net/npm/ol@v8.1.0/dist/ol.js`
 - allows style from `https://cdn.jsdelivr.net/npm/ol@v8.1.0/ol.css` for the openlayer map
 - allows styles from `https://cdn.hypothes.is/hypothesis/1.1524.0/build/styles/annotator.css` and `https://cdn.hypothes.is/hypothesis/1.1524.0/build/styles/highlights.css` for the proper display of the hypothesis viewer box
 - removes no longer use `panguin.genomics.cn` domain
 - updates loading resources from `http://gigadb.org` to `https://gigadb.org`

In the production gigadb websites, the hypothesis viewer box should be displayed properly, and the external scripts for the mapbrowse should be executed.

Therefore, I would like to bring back the fix for #1944 again in this PR.

Please have a look again.

kencho51 committed:
refactor: make sure to get aws and wasabi secrets

pli888 commented:
> Therefore, I would like to bring back the fix for https://github.com/gigascience/gigadb-website/issues/1944 again in this PR.

@kencho51

**issue**: After deploying the current code in the this PR onto my personal staging server, I still see the hypothesis box at the bottom of the home page:

<img width="1712" alt="Screenshot 2024-10-15 at 4 47 03 PM" src="https://github.com/user-attachments/assets/5cd8911b-644f-45c8-8c80-a433efbe9356">

When I empty my web browser cache, I still see the hypothesis box. The only way I stopped the hypothesis box from being displayed on my staging server was to delete the `
refactor: update CSP to display hypotheis and mapbrowse properly` commit from this PR.

**question**: Is there a reason to include a fix for #1944 in this PR which already contains a major infrastructure change?
**thought**: I would provide a fix for #1944 in a separate PR so this current PR can be approved.

kencho51 committed:
Revert "refactor: update CSP to display hypotheis and mapbrowse properly"

This reverts commit e149271f76a5aecab33e09a0d6312e40db746cc1.

kencho51 committed:
Revert "doc: update change log for CSP"

This reverts commit 60d1d979d48678eb058ff8b5dba385fb08a5ee28.

pli888 reviewed (approved):
**praise**: All tests are passing on my CI pipeline including the acceptance tests in `AdminDatasetForm.feature`.
**praise**: The strange annotations box is not displayed anymore  on the GigaDB home page on my staging server.

Happy to approve this PR now.
